src/commands/utils.rs

```diff
@@ -62,3 +71,93 @@ where
+///
+/// `change_txo` is the change output to add if needed (with any value).
+///
+/// `feerate_vb` is the minimum feerate (in sats/vb).
+///
+/// `min_fee` is the minimum fee (in sats) that the selection must have.
+///
+/// `max_sat_weight` is the maximum size difference (in vb) of
+/// an input in the transaction before and after satisfaction.
+pub fn select_coins_for_spend(
+    candidate_coins: &[CandidateCoin],
+    base_tx: bitcoin::Transaction,
+    change_txo: bitcoin::TxOut,
+    feerate_vb: u64,
+    min_fee: u64,
+    max_sat_weight: usize,
+) -> Result<(Vec<Coin>, bitcoin::Amount), InsufficientFunds> {
+    let max_input_weight = TXIN_BASE_WEIGHT + max_sat_weight as u32;
+    let candidates: Vec<Candidate> = candidate_coins
+        .iter()
+        .map(|cand| Candidate {
+            input_count: 1,
+            value: cand.coin.amount.to_sat(),
+            weight: max_input_weight,
+            is_segwit: true, // We only support receiving on Segwit scripts.
+        })
+        .collect();
+    // Transaction base weight is calculated from transaction with no inputs and no change output.
+    let base_weight = base_tx.weight().to_wu() as u32;
+    let long_term_feerate = FeeRate::from_sat_per_vb(LONG_TERM_FEERATE_VB as f32);
+    let target = Target {
+        value: base_tx.output.iter().map(|o| o.value).sum(),
+        feerate: FeeRate::from_sat_per_vb(feerate_vb as f32),
+        min_fee,
+    };
+    let drain_weights = DrainWeights {
+        output_weight: {
+            let mut tx_with_change = base_tx;
+            tx_with_change.output.push(change_txo);
+            tx_with_change.weight().to_wu() as u32 - base_weight
+        },
+        spend_weight: max_input_weight,
+    };
+    // Change policy ensures any change output is not too small and that transaction waste is reduced.
+    let change_policy =
+        change_policy::min_value_and_waste(drain_weights, DUST_OUTPUT_SATS, long_term_feerate);
+
+    let mut selector = CoinSelector::new(&candidates, base_weight);
+    // Select any mandatory candidates.
+    for (i, cand) in candidate_coins.iter().enumerate() {
+        if cand.must_select {
+            selector.select(i);
+        }
+    }
+    if let Err(e) = selector.run_bnb(
+        LowestFee {
+            target,
+            long_term_feerate,
+            change_policy: &change_policy,
+        },
+        100_000,
+    ) {
+        warn!(
+            "Coin selection error: '{}'. Selecting coins by descending value per weight unit...",
+            e.to_string()
+        );
+        // If selection not possible by BnB, order coins by descending value ready for selection below.
+        selector.sort_candidates_by_descending_value_pwu();
+    }
+    selector.select_until_target_met(target, change_policy(&selector, target))?;
+    let drain = change_policy(&selector, target);
+    let change_amount = bitcoin::Amount::from_sat(drain.value);
```

Comment:
As we'll be keeping the fallback, we'll need something like `select_until_target_met()`, but without fixing the change at the start. I think the following could work:
```
while !selector.is_target_met(target, change_policy(&selector, target)) {
        if !selector.select_next() {
            let missing = selector
                .excess(target, change_policy(&selector, target))
                .unsigned_abs();
            return Err(InsufficientFunds { missing });
        }
    }
```
We need to pass `change_policy(&selector, target)` each time we select a new candidate so that we get the right amount of change for the selection. With `select_until_target_met()`, the change amount stays the same even as additional candidates are selected.

Reply:
Not necessary to change, i was just trying to understand how comes we'd pass it twice.

Reply:
It was good you asked as in passing it made me realise that `select_until_target_met()` was not working quite how I thought it was :relieved: I think we will need to change that, although it will only affect the fallback case.

Reply:
I think after all that no changes are needed here.

Assuming that no BnB solution has been found, then `change_policy(&selector, target)` should give `Drain::none()`, i.e. no change output, and calling `select_until_target_met(target, Drain::none())` will keep selecting coins until we cover the target without any need for change, and then `let drain = change_policy(&selector, target);` will tell us how much change should be given for this final selection.

In the case that BnB has found a solution, then calling `select_until_target_met()` doesn't actually do anything and we could skip to the next line. But this way covers both cases.

I should mention that this approach with the fallback is based on https://github.com/evanlinjin/bdk/blob/fce13ac51571dcdeace2e6092b14c77aca91db10/example-crates/example_cli/src/lib.rs#L294-L329.

Reply:
Can you explain this in a comment in the next push?

Reply:
I've added a comment and also moved `select_until_target_met()` to within the fallback case only, as we can assume if BnB was successful that the required coins to meet the target have already been selected.

I think something like the following, which would cover both BnB and fallback cases, might be clearer and help avoid doubts about the use of `select_until_target_met()`:
```
let drain = loop {
        let drain = change_policy(&selector, target);
        if selector.is_target_met(target, drain) {
            break drain
        }
        if !selector.select_next() {
            let missing = selector.excess(target, drain).unsigned_abs();
            return Err(InsufficientFunds { missing });
        }
    };
```

Reply:
I think the current code with the comment is fine. Thanks!